
# World Population Collection
To review the United Nations (UN) population data and economic indicators at the country and geographic level, this notebook will be used to collect and wrangle different UN dataset and other organization's datasets The collected and wranged result results will be stored in ../World_Population_dataset for future research and evaluation.

This collection is broken into two sections:
* World Population Collections from UN
* Interesting World Data
## World Population Collections from UN
The United Nations populations data groups the world by Countries, the whole World, Continents, Geographical Sections, and Economic Groups. This UN population data are segmented into three datasets: Population, Period Indicators and Mortality Indicators. These dataset will be the basic for my first and second capstone on World population.
These csv files are directly pulled from the UN and the DataFrames are saved under ../World_Population_datasets subfolder.
## Interesting World Data
This next collection of World and Country data will mainly be used for my second capstone regarding World Population. Some of these datasets were manually download excel files from World Bank, International Monitary Fund, World Health Organization, and some other world organizations. The manually downloaded files are found under ../World_Population_datasets/source subfolder. The resulting DataFrames are stored under ../World_Population_datasets subfolder for future use.

----

## Contents for this World Popluation Wrangling

[World Population Collection from UN](#UN3-main)

1. [UN World Population Dataset Triplex](#UN3-overview) <br>
    1.1 [General Header Columns in each UN Worl Population Dataset](#UN3-header)<br>
2. [Collect UN Dataset Triplex](#UN3-coll) <br>
3. [Evaluating Population CSV Dataset](#UN3-pop) <br>
    3.1 [Population CSV Dataset Information, Columns, and Samples](#UN3-pop-header)
4. [Evaluating Period Indicators](#UN3-period) <br>
    4.1 [Period Indicators CSV Dataset Information, Columns, and Samples](#UN3-period-header)
5. [Evaluating Mortality Indicators](#UN3-mort) <br>
    5.1 [Mortaility Indicators CSV Dataset Information, Columns, and Samples](#UN3-1-5_1)<br>
    
[Interesting World Data](#Inter-main)

1. [World Economic Outlook by Country from IMF](#Inter-weo-c)<br>
   

----


# World Population Collection from UN <a id="UN3-main"></a>


## Overview UN World Population Dataset Triplex (2017) <a id="UN3-overview"></a>

|table |Description |
|:------|:-----------|
|Population |Population by 5-year age groups, annually from 1950 to 2100. Note: for years below 1990, last age available is 80+ (80 and over); for 1990 and over, age groups are available until 100+ (100 and over).|
|Period Indicators |There are several indicators that are available for 5-year periods, from 1950-1955 to 2095-2100.|
|Mortaility Indicators|Abridged life tables up to age 85 by sex and both sexes combined providing a set of values showing the mortality experience of a hypothetical group of infants born at the same time and subject throughout their lifetime to the specific mortality rates of a given period, from 1950-1955 to 2095-2100.|

### General Header Columns in each UN Worl Population Dataset <a id="UN3-header"></a>
The CSV files used are encoded in UTF-8 and all of them have the following columns:

* LocID (numeric): numeric code for the location; for countries and areas, it follows the ISO 3166-1 numeric standard
* Location (string): name of the region, subregion, country or area
* VarID (numeric): numeric code for the variant
* Variant (string): projection variant name (Medium is the most used)
* Time (string): label identifying the single year (e.g. 1950) or the period of the data (e.g. 1950-1955)
* MidPeriod (numeric): numeric value identifying the mid period of the data, with the decimal representing the month (e.g. 1950.5 for July 1950)
Use the LocID, VarID and Time columns to link the data accross the different files, if necessary. Note that Time differs between single year (e.g. 1950) and period (e.g. 1950-1955) data.

For the files with data separated by age, the following columns are available:

* AgeGrp (string): label identifying the single age (e.g. 15) or age group (e.g. 15-19)
* AgeGrpStart (numeric): initial age of the age group
* AgeGrpSpan (numeric): length of the age group, in years
* Refer to each file listed below for a description of it's specific data columns.

Please note these files don't include footnotes, please refer to the Excel files for notes that apply to the regions, subregions, countries or areas listed in the CSV files.

## Collect the 3 UN Population CSV Datasets <a id="UN3-coll"></a>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

In [2]:
# Read UN Population, Period Indicators and Mortality Indicators dataset into 3 dataframes:
pop_df = pd.read_csv('https://esa.un.org/unpd/wpp/DVD/Files/1_Indicators%20(Standard)/CSV_FILES/WPP2017_PopulationByAgeSex_Medium.csv')
period_df = pd.read_csv('https://esa.un.org/unpd/wpp/DVD/Files/1_Indicators%20(Standard)/CSV_FILES/WPP2017_Period_Indicators_Medium.csv')
mortality_df = pd.read_csv('https://esa.un.org/unpd/wpp/DVD/Files/1_Indicators%20(Standard)/CSV_FILES/WPP2017_LifeTable.csv')

## Evaluating Population CSV Dataset <a id="UN3-pop"></a>
After the 9 header columns, PopMale, PopFemale, and PopTotal are the three unique columns for this dataset. Each are all populated. As expected, some of the older age groups (95 & 100+) have some zero values, espeically in some very small countries. Even so, the UN Polulation CSV dataset is fully populated.

Note 

|LocID value|Location types|
|:---|:---|
| < 900|Countries|
| = 900|the World|
| > 900|Continents, Geographical Sections, Economic Groups, and Devlopmental Groups|

In [3]:
pop_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 725651 entries, 0 to 725650
Data columns (total 12 columns):
LocID          725651 non-null int64
Location       725651 non-null object
VarID          725651 non-null int64
Variant        725651 non-null object
Time           725651 non-null int64
MidPeriod      725651 non-null float64
AgeGrp         725651 non-null object
AgeGrpStart    725651 non-null int64
AgeGrpSpan     725651 non-null int64
PopMale        725651 non-null float64
PopFemale      725651 non-null float64
PopTotal       725651 non-null float64
dtypes: float64(4), int64(5), object(3)
memory usage: 66.4+ MB


Population Dataset does not seem to have any null values. We will check the values.

In [4]:
pop_df.describe()

LocID     VarID           Time      MidPeriod    AgeGrpStart  \
count  725651.000000  725651.0  725651.000000  725651.000000  725651.000000   
mean      563.705394       2.0    2027.949186    2028.449186      47.741614   
std       552.476813       0.0      42.679121      42.679121      29.370663   
min         4.000000       2.0    1950.000000    1950.500000       0.000000   
25%       258.000000       2.0    1993.000000    1993.500000      20.000000   
50%       504.000000       2.0    2029.000000    2029.500000      45.000000   
75%       780.000000       2.0    2065.000000    2065.500000      70.000000   
max      5501.000000       2.0    2100.000000    2100.500000     100.000000   

          AgeGrpSpan        PopMale      PopFemale       PopTotal  
count  725651.000000  725651.000000  725651.000000  725651.000000  
mean        4.699103    6906.535707    6824.501280   13731.037014  
std         1.309520   30250.119516   29230.709413   59461.110624  
min        -1.000000       0.000000       0.000000       0.000000  
25%         5.000000      18.941000      21.889000      41.679000  
50%         5.000000     173.652000     184.107000     359.645000  
75%         5.000000    1230.060000    1276.621000    2509.709500  
max         5.000000  359751.415000  342147.155000  701899.649000

All looks okay, except for minimal value of zero (0) for PopMale, PopFemale, and PopTotal.

In [5]:
pop_df[(pop_df.PopMale==0) | (pop_df.PopFemale==0) | (pop_df.PopTotal==0)].head()

LocID             Location  VarID Variant  Time  MidPeriod AgeGrp  \
15754     28  Antigua and Barbuda      2  Medium  1990     1990.5  95-99   
15755     28  Antigua and Barbuda      2  Medium  1990     1990.5   100+   
15775     28  Antigua and Barbuda      2  Medium  1991     1991.5  95-99   
15776     28  Antigua and Barbuda      2  Medium  1991     1991.5   100+   
15797     28  Antigua and Barbuda      2  Medium  1992     1992.5   100+   

       AgeGrpStart  AgeGrpSpan  PopMale  PopFemale  PopTotal  
15754           95           5      0.0      0.009     0.009  
15755          100          -1      0.0      0.001     0.001  
15775           95           5      0.0      0.010     0.010  
15776          100          -1      0.0      0.001     0.001  
15797          100          -1      0.0      0.001     0.001

In [6]:
pop_df[(pop_df.PopMale==0) | (pop_df.PopFemale==0) | (pop_df.PopTotal==0)].describe()

LocID   VarID         Time    MidPeriod  AgeGrpStart  \
count  1321.000000  1321.0  1321.000000  1321.000000  1321.000000   
mean    491.093868     2.0  2008.427706  2008.927706    99.803179   
std     235.873362     0.0    16.915730    16.915730     0.972667   
min      28.000000     2.0  1990.000000  1990.500000    95.000000   
25%     296.000000     2.0  1996.000000  1996.500000   100.000000   
50%     533.000000     2.0  2004.000000  2004.500000   100.000000   
75%     678.000000     2.0  2016.000000  2016.500000   100.000000   
max     957.000000     2.0  2077.000000  2077.500000   100.000000   

        AgeGrpSpan      PopMale    PopFemale     PopTotal  
count  1321.000000  1321.000000  1321.000000  1321.000000  
mean     -0.763815     0.000005     0.000894     0.000957  
std       1.167200     0.000078     0.001117     0.001223  
min      -1.000000     0.000000     0.000000     0.000000  
25%      -1.000000     0.000000     0.000000     0.000000  
50%      -1.000000     0.000000     0.001000     0.001000  
75%      -1.000000     0.000000     0.001000     0.001000  
max       5.000000     0.002000     0.011000     0.012000

Okay, ages of 95 and above in small countries have one to a dozen people that are either all Females or all Males, or nobody. That makes sense, they collected the data and zero was the answer for the age group. The Population Dataset does not have any null values or zero values that should be null.

### Population CSV Dataset Information, Columns, and Samples <a id="UN3-pop-header"></a>
Population interpolated by single age, annually from 1950 to 2100. 
Note: for years below 1990, last age available is 80+ (80 and over); for 1990 and over, individual ages are available until 100+ (100 and over).

|Fields|Descriptions|
|:---|:---|
|PopMale| Male population in the age group (thousands)|
|PopFemale| Female population in the age group (thousands)|
|PopTotal| Total population in the age group (thousands)|


In [7]:
pop_df.head(20)

LocID     Location  VarID Variant  Time  MidPeriod AgeGrp  AgeGrpStart  \
0       4  Afghanistan      2  Medium  1950     1950.5    0-4            0   
1       4  Afghanistan      2  Medium  1950     1950.5    5-9            5   
2       4  Afghanistan      2  Medium  1950     1950.5  10-14           10   
3       4  Afghanistan      2  Medium  1950     1950.5  15-19           15   
4       4  Afghanistan      2  Medium  1950     1950.5  20-24           20   
5       4  Afghanistan      2  Medium  1950     1950.5  25-29           25   
6       4  Afghanistan      2  Medium  1950     1950.5  30-34           30   
7       4  Afghanistan      2  Medium  1950     1950.5  35-39           35   
8       4  Afghanistan      2  Medium  1950     1950.5  40-44           40   
9       4  Afghanistan      2  Medium  1950     1950.5  45-49           45   
10      4  Afghanistan      2  Medium  1950     1950.5  50-54           50   
11      4  Afghanistan      2  Medium  1950     1950.5  55-59           55   
12      4  Afghanistan      2  Medium  1950     1950.5  60-64           60   
13      4  Afghanistan      2  Medium  1950     1950.5  65-69           65   
14      4  Afghanistan      2  Medium  1950     1950.5  70-74           70   
15      4  Afghanistan      2  Medium  1950     1950.5  75-79           75   
16      4  Afghanistan      2  Medium  1950     1950.5    80+           80   
17      4  Afghanistan      2  Medium  1951     1951.5    0-4            0   
18      4  Afghanistan      2  Medium  1951     1951.5    5-9            5   
19      4  Afghanistan      2  Medium  1951     1951.5  10-14           10   

    AgeGrpSpan  PopMale  PopFemale  PopTotal  
0            5  630.044    661.578  1291.622  
1            5  516.206    487.335  1003.540  
2            5  461.378    423.326   884.704  
3            5  414.369    369.363   783.732  
4            5  374.109    318.392   692.501  
5            5  321.311    272.299   593.611  
6            5  276.279    232.168   508.446  
7            5  236.792    197.326   434.120  
8            5  200.616    167.179   367.796  
9            5  176.525    141.220   317.745  
10           5  147.126    117.700   264.826  
11           5  123.896     95.305   219.200  
12           5   94.572     72.748   167.321  
13           5   65.057     50.043   115.104  
14           5   37.983     29.217    67.199  
15           5   17.012     13.085    30.096  
16          -1    5.968      4.590    10.555  
17           5  641.074    673.084  1314.241  
18           5  525.299    502.420  1027.747  
19           5  465.584    427.584   893.176

In [8]:
# Loction Examples of Continents, Geographical Sections, and Economic Groups 
# Countries have LocID < 900
pop_df[pop_df['LocID']>=900].groupby('Location')['LocID'].mean()

Location
Africa                                                          903
Asia                                                            935
Australia/New Zealand                                           927
Caribbean                                                       915
Central America                                                 916
Central Asia                                                   5500
Eastern Africa                                                  910
Eastern Asia                                                    906
Eastern Europe                                                  923
Europe                                                          908
High-income countries                                          1503
Latin America and the Caribbean                                 904
Least developed countries                                       941
Less developed regions                                          902
Less developed regions, excluding China

## Evaluating Period Indicators <a id="UN3-period"></a>
32 countries only have Growth Rate data. The other Location have all the Period Indicator data populated, including Growth Rate. The segratation of the Period Indicator dataset by Growth Rate versus no null rows will be left to evaluation and usage stages.

In [9]:
period_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8190 entries, 0 to 8189
Data columns (total 25 columns):
LocID            8190 non-null int64
Location         8190 non-null object
VarID            8190 non-null int64
Variant          8190 non-null object
Time             8190 non-null object
MidPeriod        8190 non-null float64
TFR              7230 non-null float64
NRR              7230 non-null float64
CBR              7230 non-null float64
Births           7230 non-null float64
LEx              7230 non-null float64
LExMale          7230 non-null float64
LExFemale        7230 non-null float64
IMR              7230 non-null float64
Q5               7230 non-null float64
CDR              7230 non-null float64
Deaths           7230 non-null float64
DeathsMale       7230 non-null float64
DeathsFemale     7230 non-null float64
CNMR             7230 non-null float64
NetMigrations    7230 non-null float64
GrowthRate       8190 non-null float64
NatIncr          7230 non-null float64
SRB

The null entries are for island and small countries were data was not available. The list of countries are shown below, near the end of this Period Indicators section.

### Period Indicators CSV Dataset Information, Columns, and Samples <a id="UN3-period-header"></a>
Several indicators that are available for 5-year periods, from 1950-1955 to 2095-2100. Note that there will be 30 entries from 1950-1955 to 2095-2100.

|Fields|Descriptions|
|:---|:---|
|TFR| Total fertility (live births per woman)
|NRR| Net reproduction rate (surviving daughters per woman)
|CBR| Crude birth rate (births per 1,000 population)
|Births| Number of births, both sexes combined (thousands)
|LEx *|Life expectancy at birth for both sexes combined (years)
|LExMale *|Male life expectancy at birth (years)
|LExFemale *|Female life expectancy at birth (years)
|IMR *|Infant mortality rate, q(1), for both sexes combined (infant deaths per 1,000 live births)
|Q5 *| Under-five mortality, 5q0, for both sexes combined (deaths under age five per 1,000 live births)
|CDR| Crude death rate (deaths per 1,000 population)
|Deaths | Number of deaths, both sexes combined (thousands)
|DeathsMale| Number of male deaths (thousands)
|DeathsFemale| Number of female deaths (thousands)
|CNMR *| Net migration rate (per 1,000 population)
|NetMigrations *| Net number of migrants, both sexes combined (thousands)
|GrowthRate| Average annual rate of population change (percentage)
|NatIncr| Rate of natural increase (per 1,000 population)
|SRB *| Sex ratio at birth (male births per female births)
|MAC *| Female mean age of childbearing (years)
\* not published for variants other than Medium.

In [10]:
period_df.head(10)

LocID     Location  VarID Variant       Time  MidPeriod    TFR    NRR  \
0      4  Afghanistan      2  Medium  1950-1955     1953.0  7.450  1.636   
1      4  Afghanistan      2  Medium  1955-1960     1958.0  7.450  1.765   
2      4  Afghanistan      2  Medium  1960-1965     1963.0  7.450  1.880   
3      4  Afghanistan      2  Medium  1965-1970     1968.0  7.450  1.986   
4      4  Afghanistan      2  Medium  1970-1975     1973.0  7.450  2.094   
5      4  Afghanistan      2  Medium  1975-1980     1978.0  7.450  2.213   
6      4  Afghanistan      2  Medium  1980-1985     1983.0  7.450  2.357   
7      4  Afghanistan      2  Medium  1985-1990     1988.0  7.469  2.551   
8      4  Afghanistan      2  Medium  1990-1995     1993.0  7.482  2.737   
9      4  Afghanistan      2  Medium  1995-2000     1998.0  7.654  2.912   

      CBR    Births   ...       CDR    Deaths  DeathsMale  DeathsFemale  \
0  50.316  2015.476   ...    36.873  1477.014     797.575       679.439   
1  51.001  2201.589   ...    33.725  1455.821     782.293       673.528   
2  51.513  2438.489   ...    31.190  1476.423     787.735       688.688   
3  51.754  2725.445   ...    28.821  1517.733     804.151       713.582   
4  51.415  3048.468   ...    26.384  1564.313     824.381       739.932   
5  50.840  3284.102   ...    23.740  1533.549     805.681       727.868   
6  50.332  3149.701   ...    20.722  1296.726     680.548       616.178   
7  49.650  2982.985   ...    17.189  1032.727     542.993       489.734   
8  48.668  3570.874   ...    14.282  1047.923     556.021       491.902   
9  49.139  4569.075   ...    12.856  1195.382     635.117       560.265   

     CNMR  NetMigrations  GrowthRate  NatIncr   SRB     MAC  
0  -0.499        -20.000       1.295   13.443  1.06  29.835  
1  -0.463        -20.000       1.682   17.276  1.06  29.835  
2  -0.423        -20.000       1.992   20.323  1.06  29.835  
3  -0.380        -20.000       2.258   22.933  1.06  29.835  
4  -0.337        -20.000       2.473   25.031  1.06  29.835  
5 -16.912      -1092.462       1.019   27.100  1.06  29.871  
6 -53.026      -3318.301      -2.344   29.610  1.06  29.904  
7 -24.703      -1484.185       0.776   32.461  1.06  29.917  
8  31.722       2327.473       6.672   34.386  1.06  29.884  
9  -4.081       -379.474       3.227   36.283  1.06  29.779  

[10 rows x 25 columns]

In [11]:
x = period_df.Location.value_counts(dropna=False).max()
n = period_df.Location.value_counts(dropna=False).min()
print("The maximum and minimum value for location count of Period Indicators locations are ",x,"and",n,".")

The maximum and minimum value for location count of Period Indicators locations are  30 and 30 .


#### Proof that Period Indicators rows with null values are all isolated to certain locations
We are going to show that a small list of locations have all the null values. From the info() on Period Indicators DataFrame above, some locations only have Growth Rate data (the first 6 head columns define the countries' rows).  Looking rows with null values for TRF (below), show a few examples where only the row header columns and Growth Rate have non-null value.  

To prove this we will set the null hypothesis $H_0$ = For the 18 columns with 960 nulls each, they are not in all of 30 rows for each of the 32 locations, and do not account for exactly 960 rows total (30 * 32 = 960). 
The opposite of null hypothesis is $H_1$ = there are exactly 32 locations with all of their 30 rows containing the null values and total 960 rows.

So first off, here is the example of 5 rows from American Samoa with everything but the header columns and Growth Rate set to null.

In [12]:
period_df[period_df.TFR.isnull()].head()

LocID        Location  VarID Variant       Time  MidPeriod  TFR  NRR  \
120     16  American Samoa      2  Medium  1950-1955     1953.0  NaN  NaN   
121     16  American Samoa      2  Medium  1955-1960     1958.0  NaN  NaN   
122     16  American Samoa      2  Medium  1960-1965     1963.0  NaN  NaN   
123     16  American Samoa      2  Medium  1965-1970     1968.0  NaN  NaN   
124     16  American Samoa      2  Medium  1970-1975     1973.0  NaN  NaN   

     CBR  Births ...   CDR  Deaths  DeathsMale  DeathsFemale  CNMR  \
120  NaN     NaN ...   NaN     NaN         NaN           NaN   NaN   
121  NaN     NaN ...   NaN     NaN         NaN           NaN   NaN   
122  NaN     NaN ...   NaN     NaN         NaN           NaN   NaN   
123  NaN     NaN ...   NaN     NaN         NaN           NaN   NaN   
124  NaN     NaN ...   NaN     NaN         NaN           NaN   NaN   

     NetMigrations  GrowthRate  NatIncr  SRB  MAC  
120            NaN       0.796      NaN  NaN  NaN  
121            NaN       0.308      NaN  NaN  NaN  
122            NaN       3.229      NaN  NaN  NaN  
123            NaN       2.977      NaN  NaN  NaN  
124            NaN       1.847      NaN  NaN  NaN  

[5 rows x 25 columns]

Now, we will first show that the nulls for TFR are isolated to the 30 rows for 32 locations in the next two queries. The first query below searchs for only the rows where TFR are null, and counts up how many are there for each location found. The second query just repeats the first query and counts the rows.

In [13]:
period_df[period_df.TFR.isnull()].Location.value_counts(dropna=False)

Sint Maarten (Dutch part)      30
Marshall Islands               30
Greenland                      30
Cayman Islands                 30
San Marino                     30
Tokelau                        30
Caribbean Netherlands          30
Saint Kitts and Nevis          30
Falkland Islands (Malvinas)    30
Bermuda                        30
Liechtenstein                  30
Saint Pierre and Miquelon      30
Anguilla                       30
Dominica                       30
Holy See                       30
Isle of Man                    30
Tuvalu                         30
Monaco                         30
Saint Helena                   30
Montserrat                     30
Gibraltar                      30
Wallis and Futuna Islands      30
Faeroe Islands                 30
Palau                          30
Nauru                          30
Andorra                        30
Cook Islands                   30
Northern Mariana Islands       30
American Samoa                 30
British Virgin

In [14]:
period_df[period_df.TFR.isnull()].Location.value_counts(dropna=False).count()

32

From the two searchs above, we can see the TFR 960 nulls are only in 32 location with each location having 30 of its 30 rows having null TFR values. Now our null hypothesis $H_0$ can still be true if TFR is the only column with 32 location with all 30 rows set to null. This can easily be shown by removing all the rows with any column having a null and showing that there are more than 960 rows removed. With Period Indicator DataFrame having 8190 rows, we need to have fewer than 7230 rows (8190 - 960 = 7230) to show that more than 32 countries with 30 rows have null and that $H_0$ is true.

In [15]:
temp_df= period_df.dropna(axis=0,how='any')
temp_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7230 entries, 0 to 8189
Data columns (total 25 columns):
LocID            7230 non-null int64
Location         7230 non-null object
VarID            7230 non-null int64
Variant          7230 non-null object
Time             7230 non-null object
MidPeriod        7230 non-null float64
TFR              7230 non-null float64
NRR              7230 non-null float64
CBR              7230 non-null float64
Births           7230 non-null float64
LEx              7230 non-null float64
LExMale          7230 non-null float64
LExFemale        7230 non-null float64
IMR              7230 non-null float64
Q5               7230 non-null float64
CDR              7230 non-null float64
Deaths           7230 non-null float64
DeathsMale       7230 non-null float64
DeathsFemale     7230 non-null float64
CNMR             7230 non-null float64
NetMigrations    7230 non-null float64
GrowthRate       7230 non-null float64
NatIncr          7230 non-null float64
SRB

After removing any row with one or more nulls, there are 7230 rows, which is a decrease of 960 rows. This exactly conisides with what was seen with TFR, is not less than 7230 rows, and shows that $H_0$ is false.

So, $H_1$ is true. 32 locations only have Growth Rate columns set in the Period Indicators DataFrame, with the rest of the variable column set to null.


## Evaluating Mortality Indicators <a id="UN3-mort"></a>
The only null values are for moratility/life probabilities at the oldest range (85+). Since 85+ includes people living to infinity, insurance tables say there is no reason to speculate on probalibilies. So, this dataset's values are good.

In [16]:
mortality_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 412110 entries, 0 to 412109
Data columns (total 21 columns):
LocID          412110 non-null int64
Location       412110 non-null object
VarID          412110 non-null int64
Variant        412110 non-null object
Time           412110 non-null object
MidPeriod      412110 non-null float64
SexID          412110 non-null int64
Sex            412110 non-null object
AgeGrp         412110 non-null object
AgeGrpStart    412110 non-null int64
AgeGrpSpan     412110 non-null int64
mx             412110 non-null float64
qx             390420 non-null float64
px             390420 non-null float64
lx             412110 non-null float64
dx             412110 non-null float64
Lx             412110 non-null float64
Sx             390420 non-null float64
Tx             412110 non-null float64
ex             412110 non-null float64
ax             412110 non-null float64
dtypes: float64(11), int64(5), object(5)
memory usage: 66.0+ MB


### Mortaility Indicators CSV Dataset Information, Columns, and Samples <a id="UN3-mort-header></a>
Abridged life tables up to age 85 by sex and both sexes combined providing a set of values showing the mortality experience of a hypothetical group of infants born at the same time and subject throughout their lifetime to the specific mortality rates of a given period, from 1950-1955 to 2095-2100.

|Fields|Descriptions|
|:---|:---|
|mx| Central death rate, nmx, for the age interval (x, x+n)|
|qx| Probability of dying (nqx), for an individual between age x and x+n|
|px| Probability of surviving, (npx), for an individual of age x to age x+n|
|lx| Number of survivors, (lx), at age (x) for 100000 births|
|dx| Number of deaths, (ndx), between ages x and x+n|
|Lx| Number of person-years lived, (nLx), between ages x and x+n|
|Sx| Survival ratio (nSx) corresponding to proportion of the life table population in age group (x, x+n) who are alive n year later|
|Tx| Person-years lived, (Tx), above age x|
|ex| Expectation of life (ex) at age x, i.e., average number of years lived subsequent to age x by those reaching age x|
|ax| Average number of years lived (nax) between ages x and x+n by those dying in the interval|


In [17]:
mortality_df.head(10)

LocID     Location  VarID Variant       Time  MidPeriod  SexID   Sex  \
0      4  Afghanistan      2  Medium  1950-1955     1953.0      1  Male   
1      4  Afghanistan      2  Medium  1950-1955     1953.0      1  Male   
2      4  Afghanistan      2  Medium  1950-1955     1953.0      1  Male   
3      4  Afghanistan      2  Medium  1950-1955     1953.0      1  Male   
4      4  Afghanistan      2  Medium  1950-1955     1953.0      1  Male   
5      4  Afghanistan      2  Medium  1950-1955     1953.0      1  Male   
6      4  Afghanistan      2  Medium  1950-1955     1953.0      1  Male   
7      4  Afghanistan      2  Medium  1950-1955     1953.0      1  Male   
8      4  Afghanistan      2  Medium  1950-1955     1953.0      1  Male   
9      4  Afghanistan      2  Medium  1950-1955     1953.0      1  Male   

  AgeGrp  AgeGrpStart  ...       mx     qx     px          lx         dx  \
0      0            0  ...    0.367  0.294  0.706  100000.000  29428.993   
1    1-4            1  ...    0.049  0.173  0.827   70571.007  12205.007   
2    5-9            5  ...    0.010  0.046  0.954   58366.001   2705.968   
3  10-14           10  ...    0.007  0.033  0.967   55660.033   1862.386   
4  15-19           15  ...    0.009  0.045  0.955   53797.647   2446.167   
5  20-24           20  ...    0.013  0.065  0.935   51351.480   3317.085   
6  25-29           25  ...    0.015  0.072  0.928   48034.394   3456.789   
7  30-34           30  ...    0.017  0.083  0.917   44577.605   3693.845   
8  35-39           35  ...    0.020  0.097  0.903   40883.760   3970.723   
9  40-44           40  ...    0.025  0.117  0.883   36913.037   4312.180   

           Lx     Sx          Tx     ex     ax  
0   80282.575  0.660  2793334.90  27.93  0.330  
1  249965.170  0.860  2713052.30  38.44  1.352  
2  283893.070  0.964  2463087.10  42.20  2.067  
3  273607.930  0.962  2179194.00  39.15  2.481  
4  263167.640  0.945  1905586.10  35.42  2.621  
5  248700.780  0.931  1642418.50  31.98  2.571  
6  231607.870  0.923  1393717.70  29.01  2.523  
7  213762.480  0.910  1162109.80  26.07  2.530  
8  194621.840  0.893   948347.34  23.20  2.533  
9  173875.520  0.875   753725.50  20.42  2.521  

[10 rows x 21 columns]

In [18]:
mortality_df.describe()

LocID     VarID     MidPeriod          SexID    AgeGrpStart  \
count  412110.000000  412110.0  412110.00000  412110.000000  412110.000000   
mean      563.705394       2.0    2025.50000       2.000000      40.315789   
std       552.477102       0.0      43.27726       0.816498      26.895645   
min         4.000000       2.0    1953.00000       1.000000       0.000000   
25%       258.000000       2.0    1988.00000       1.000000      15.000000   
50%       504.000000       2.0    2025.50000       2.000000      40.000000   
75%       780.000000       2.0    2063.00000       3.000000      65.000000   
max      5501.000000       2.0    2098.00000       3.000000      85.000000   

          AgeGrpSpan             mx             qx             px  \
count  412110.000000  412110.000000  390420.000000  390420.000000   
mean        5.263158       0.027356       0.069605       0.930391   
std         2.467521       0.053245       0.117351       0.117359   
min         1.000000       0.000000       0.000000       0.154000   
25%         5.000000       0.001000       0.004000       0.926000   
50%         5.000000       0.005000       0.018000       0.982000   
75%         5.000000       0.024000       0.074000       0.996000   
max        15.000000       0.559000       0.846000       1.000000   

                  lx             dx            Lx             Sx  \
count  412110.000000  412110.000000  4.121100e+05  390420.000000   
mean    81698.548493    5263.158163  3.773705e+05       0.911579   
std     23311.560744    9231.340617  1.378261e+05       0.150493   
min        14.017000       1.620000  3.298800e+01       0.103000   
25%     73160.794250     432.743000  3.136663e+05       0.910000   
50%     92502.789500    1844.290500  4.308024e+05       0.979000   
75%     98330.513500    6734.937250  4.841751e+05       0.995000   
max    100000.000000   95599.630000  1.327014e+06       1.000000   

                 Tx             ex             ax  
count  4.121100e+05  412110.000000  412110.000000  
mean   3.567734e+06      38.771423       2.580589  
std    2.487559e+06      23.576486       1.054504  
min    3.298800e+01       1.790000       0.043000  
25%    1.311221e+06      17.400000       2.518000  
50%    3.328074e+06      37.020000       2.612000  
75%    5.597354e+06      58.050000       2.674000  
max    9.742800e+06      97.430000      14.647000

### Looking at Mortality Indicator null records
Below we can see that qx (Probability of dying), px (Probability of survial), and Sx (Survival ratio) are null for that oldest age range for each location and each year range. In the example below, it shows Afghanistan rage range of 80+ for Male, Female and Total rows for time range 1950-1955.

In [19]:
mortality_df[(mortality_df.px.isnull()) & (mortality_df.qx.isnull()) & (mortality_df.Sx.isnull())].head(6)

LocID     Location  VarID Variant       Time  MidPeriod  SexID     Sex  \
18       4  Afghanistan      2  Medium  1950-1955     1953.0      1    Male   
37       4  Afghanistan      2  Medium  1950-1955     1953.0      2  Female   
56       4  Afghanistan      2  Medium  1950-1955     1953.0      3   Total   
75       4  Afghanistan      2  Medium  1955-1960     1958.0      1    Male   
94       4  Afghanistan      2  Medium  1955-1960     1958.0      2  Female   
113      4  Afghanistan      2  Medium  1955-1960     1958.0      3   Total   

    AgeGrp  AgeGrpStart  ...       mx  qx  px        lx        dx        Lx  \
18     85+           85  ...    0.365 NaN NaN   525.401   525.401  1440.955   
37     85+           85  ...    0.337 NaN NaN   865.542   865.542  2567.412   
56     85+           85  ...    0.352 NaN NaN   657.045   657.045  1865.547   
75     85+           85  ...    0.350 NaN NaN   751.265   751.265  2148.706   
94     85+           85  ...    0.323 NaN NaN  1212.600  1212.600  3757.479   
113    85+           85  ...    0.337 NaN NaN   936.606   936.606  2781.012   

     Sx        Tx    ex     ax  
18  NaN  1440.955  2.74  2.743  
37  NaN  2567.412  2.97  2.966  
56  NaN  1865.547  2.84  2.839  
75  NaN  2148.706  2.86  2.860  
94  NaN  3757.479  3.10  3.099  
113 NaN  2781.012  2.97  2.969  

[6 rows x 21 columns]

Below, we find that there are 21690 rows where qx, px, and Sx all null. Sx, qx, and sx are the only three columns that have a null (from info done above). Also, each location has 90 entries: 3 sets for Male, Female and Total(both) overs the 30 year ranges. After counting the number of locations with the 3 null as 241, we multiple 90 * 241 to see if the result is less than 21690, which would mean that not all locations have px, qx, and Sx all being null. Another constraining result would be that the lowest number of rows for a Location with the 3 nulls would be less than 90 or greater than 90

In [20]:
x = mortality_df[(mortality_df.px.isnull()) & (mortality_df.qx.isnull()) & (mortality_df.Sx.isnull())].Location.count()
print("number of rows where px, qx, and Sx are all null in the same row is",x)
print("A sample of Location group counts where px, qx, and Sx are all null in the row are as follows:")
mortality_df[(mortality_df.px.isnull()) & (mortality_df.qx.isnull()) & (mortality_df.Sx.isnull())].Location.value_counts(dropna=False).head(10)

number of rows where px, qx, and Sx are all null in the same row is 21690
A sample of Location group counts where px, qx, and Sx are all null in the row are as follows:


Lao People's Democratic Republic    90
Uganda                              90
Madagascar                          90
Libya                               90
El Salvador                         90
Eastern Africa                      90
French Polynesia                    90
Comoros                             90
Montenegro                          90
Serbia                              90
Name: Location, dtype: int64

In [21]:
print("From info on Mortality Indicators: 412110 - 390420 =",412110-390420)
y = mortality_df[(mortality_df.px.isnull()) & (mortality_df.qx.isnull()) & (mortality_df.Sx.isnull())].Location.value_counts(dropna=False).max()
print("Largest number of rows for a Location with px, qx, and Sx all being null is",y)
y = mortality_df[(mortality_df.px.isnull()) & (mortality_df.qx.isnull()) & (mortality_df.Sx.isnull())].Location.value_counts(dropna=False).min()
print("Lowest number of rows for a Location with px, qx, and Sx all being null is",y)
x = mortality_df[(mortality_df.px.isnull()) & (mortality_df.qx.isnull()) & (mortality_df.Sx.isnull())].Location.value_counts(dropna=False).count()
print("Locations having the 3 nulls is",x )
print(y,"*",x, "=",y*x)



From info on Mortality Indicators: 412110 - 390420 = 21690
Largest number of rows for a Location with px, qx, and Sx all being null is 90
Lowest number of rows for a Location with px, qx, and Sx all being null is 90
Locations having the 3 nulls is 241
90 * 241 = 21690


## Saving the desired Population, Period Indicators and Mortality Indicators datasets

For Population dataset, the DataFrame did not have nulls, , and can just be saved and stored. 

For the Mortality Indicators

For the period Indicators, we might want to see is we can fill the missing data from these different countries. First

In [22]:
# saving Population, Period Indicators and Moratilty Indicators datasets
#df.to_csv(file_name, sep='\t', encoding='utf-8', index=False, sep='\t')
pop_df.to_csv('../World_Population_datasets/un_population.csv', index=False)
period_df.to_csv('../World_Population_datasets/un_period_ind.csv', index=False)
mortality_df.to_csv('../World_Population_datasets/un_mortality_ind.csv', index=False)

# Interesting World Data  <a id="Inter-main"></a>
These dataset are mostly options for the second capstone. The some of WEO values may be used for the first capstone.

## World Economic Outlook by Country from IMF <a id="Inter-weo-c"></a>
The WEO data covers 193 countries from 1980 to present. There are 24 different types of monitary data available for each country.

### Notes from IMF
Download entire World Economic Outlook database
April 2018

The entire World Economic Outlook database is available for download in Tab Delimited Values format and SDMX format. All available data that can be released to the public are incorporated into both formats.The "By Countries" file contains series data for all countries, and the "By Country Groups" file contains series data for all country groups (aggregated data).

Database version
By default, results are returned from the most recent WEO report dataset. Occasionally some countries' data are adjusted, typically when index years or currencies change.

###IMF WEO April 2018
 website link found on http://www.imf.org/external/pubs/ft/weo/2018/01/weodata/download.aspx  
http://www.imf.org/external/pubs/ft/weo/2018/01/weodata/WEOApr2018all.xls

In [23]:
# IMF International monitary fund & World Economic Outlook (WEO)
# Select Country Group (all) mnaually created and downloaded weoreptc.aspx (tab delimited)
# https://www.imf.org/external/pubs/ft/weo/2017/02/weodata/weoselgr.aspx
#2017
imf_df = pd.read_csv('../World_Population_datasets/source/imf_weoreptc.aspx', encoding='iso-8859-15', sep='\t')


In [24]:
imf_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4633 entries, 0 to 4632
Data columns (total 50 columns):
WEO Country Code                 4633 non-null object
Country                          4632 non-null object
Subject Descriptor               4632 non-null object
Units                            4632 non-null object
Scale                            965 non-null object
Country/Series-specific Notes    3895 non-null object
1980                             2117 non-null object
1981                             2167 non-null object
1982                             2186 non-null object
1983                             2209 non-null object
1984                             2224 non-null object
1985                             2263 non-null object
1986                             2281 non-null object
1987                             2295 non-null object
1988                             2337 non-null object
1989                             2363 non-null object
1990                          

In [39]:
imf_df.groupby('Country')['Subject Descriptor'].count().count()

193

In [40]:
imf_df.groupby('Country')['Subject Descriptor'].count()

Country
Afghanistan                 24
Albania                     24
Algeria                     24
Angola                      24
Antigua and Barbuda         24
Argentina                   24
Armenia                     24
Australia                   24
Austria                     24
Azerbaijan                  24
Bahrain                     24
Bangladesh                  24
Barbados                    24
Belarus                     24
Belgium                     24
Belize                      24
Benin                       24
Bhutan                      24
Bolivia                     24
Bosnia and Herzegovina      24
Botswana                    24
Brazil                      24
Brunei Darussalam           24
Bulgaria                    24
Burkina Faso                24
Burundi                     24
Cabo Verde                  24
Cambodia                    24
Cameroon                    24
Canada                      24
                            ..
Switzerland                 24


In [41]:
imf_df.groupby('Subject Descriptor').Country.count().count()

21

In [36]:
imf_df.groupby('Subject Descriptor').Country.count()

Subject Descriptor
Current account balance                                                               386
Employment                                                                            193
General government gross debt                                                         193
General government net debt                                                           193
General government net lending/borrowing                                              193
General government primary net lending/borrowing                                      193
General government revenue                                                            193
General government structural balance                                                 193
General government total expenditure                                                  193
Gross domestic product based on purchasing-power-parity (PPP) share of world total    193
Gross domestic product per capita, current prices                                

In [37]:
imf_df.head(22)

WEO Country Code      Country  \
0               512  Afghanistan   
1               512  Afghanistan   
2               512  Afghanistan   
3               512  Afghanistan   
4               512  Afghanistan   
5               512  Afghanistan   
6               512  Afghanistan   
7               512  Afghanistan   
8               512  Afghanistan   
9               512  Afghanistan   
10              512  Afghanistan   
11              512  Afghanistan   
12              512  Afghanistan   
13              512  Afghanistan   
14              512  Afghanistan   
15              512  Afghanistan   
16              512  Afghanistan   
17              512  Afghanistan   
18              512  Afghanistan   
19              512  Afghanistan   
20              512  Afghanistan   
21              512  Afghanistan   

                                   Subject Descriptor  \
0              Gross domestic product, current prices   
1              Gross domestic product, current prices   
2   Gross domestic product per capita, current prices   
3              Output gap in percent of potential GDP   
4   Gross domestic product based on purchasing-pow...   
5                         Implied PPP conversion rate   
6                              Gross national savings   
7                  Inflation, average consumer prices   
8                  Inflation, average consumer prices   
9             Volume of imports of goods and services   
10                         Volume of Imports of goods   
11            Volume of exports of goods and services   
12                         Volume of exports of goods   
13                                  Unemployment rate   
14                                         Employment   
15                         General government revenue   
16               General government total expenditure   
17           General government net lending/borrowing   
18              General government structural balance   
19   General government primary net lending/borrowing   
20                        General government net debt   
21                      General government gross debt   

                                                Units     Scale  \
0                                        U.S. dollars  Billions   
1      Purchasing power parity; international dollars  Billions   
2                                        U.S. dollars     Units   
3                            Percent of potential GDP       NaN   
4                                             Percent       NaN   
5   National currency per current international do...       NaN   
6                                      Percent of GDP       NaN   
7                                               Index       NaN   
8                                      Percent change       NaN   
9                                      Percent change       NaN   
10                                     Percent change       NaN   
11                                     Percent change       NaN   
12                                     Percent change       NaN   
13                       Percent of total labor force       NaN   
14                                            Persons  Millions   
15                                     Percent of GDP       NaN   
16                                     Percent of GDP       NaN   
17                                     Percent of GDP       NaN   
18                           Percent of potential GDP       NaN   
19                                     Percent of GDP       NaN   
20                                     Percent of GDP       NaN   
21                                     Percent of GDP       NaN   

                        Country/Series-specific Notes 1980 1981 1982 1983  \
0   See notes for:  Gross domestic product, curren...  NaN  NaN  NaN  NaN   
1   See notes for:  Gross domestic product, curren...  NaN  NaN  NaN  NaN   
2   See notes for:  Gross domestic product, curren...  NaN  NaN  NaN  NaN   
3           

In [26]:
imf_df.describe()


Estimates Start After
count            3890.000000
mean             2013.206170
std                64.621553
min                 0.000000
25%              2015.000000
50%              2016.000000
75%              2016.000000
max              2017.000000

In [42]:
#popHi_df = pd.read_csv('https://esa.un.org/unpd/wpp/DVD/Files/1_Indicators%20(Standard)/CSV_FILES/WPP2017_PopulationBySingleAgeSex.csv')
##Location list with codes (numerical and ISO3), description, major area, region development group, and income group.

# GDP per capita, PPP (current international $) (2016 https://data.worldbank.org/indicator/NY.GDP.PCAP.PP.CD)
#??GDPpCap = pd.read_csv('http://api.worldbank.org/v2/en/indicator/NY.GDP.PCAP.PP.CD?downloadformat=csv')

# World Health Org (WHO) 
# manually download from http://apps.who.int/nha/database/Select/Indicators/en
# CSV excel "Who All Health Country Data"  WHO_All_Health_Country_Data

# GDP per Capita
#manually downloaded Excel https://www.clio-infra.eu/Indicators/GDPperCapita.html
# GDPperCapita_Compact.xlms



#WEO
#World Economic and Financial Surveys
#World Economic Outlook Database
# http://www.imf.org/external/pubs/ft/weo/2013/01/weodata/index.
# tab seperated files
# url_2014 = 'http://www.imf.org/external/pubs/ft/weo/2014/01/weodata/WEOApr2014all.xls'
# url_2015 = 'http://www.imf.org/external/pubs/ft/weo/2015/01/weodata/WEOApr2015all.xls'

# download from http://www.imf.org/external/pubs/ft/weo/2013/01/weodata/WEOApr2013all.xls
    
#IMF WEO April 2018
# website link found on http://www.imf.org/external/pubs/ft/weo/2018/01/weodata/download.aspx
#weo_df = pd.read_excel('http://www.imf.org/external/pubs/ft/weo/2018/01/weodata/WEOApr2018all.xls') 
# https://www.imf.org/external/pubs/ft/weo/2017/02/weodata/download.aspx
weo_df = pd.read_csv('http://www.imf.org/external/pubs/ft/weo/2018/01/weodata/WEOApr2018all.xls', sep='\t')
weo_db_df= pd.read_csv('http://www.imf.org/external/pubs/ft/weo/2018/01/weodata/WEOApr2018all.xls', encoding='iso-8859-15', sep='\t')

#, encoding='iso-8859-15', sep='\t')

## http://apps.who.int/gho/athena/data/xmart.csv?target=GHO/NUTRITION_HA_2&profile=crosstable&filter=COUNTRY:*;REGION:*;SEX:*&x-sideaxis=COUNTRY;YEAR;DATASOURCE&x-topaxis=GHO;SEX

In [ ]:
#

In [28]:
weo_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 0 entries
Data columns (total 1 columns):
<html><head><title>Request Rejected</title></head><body>The requested URL was rejected. Please consult with your administrator.<br><br>Your support ID is: 15560383175076000099</body></html>    0 non-null object
dtypes: object(1)
memory usage: 0.0+ bytes


In [43]:
weo_db_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 0 entries
Data columns (total 1 columns):
<html><head><title>Request Rejected</title></head><body>The requested URL was rejected. Please consult with your administrator.<br><br>Your support ID is: 15560383175072289430</body></html>    0 non-null object
dtypes: object(1)
memory usage: 0.0+ bytes
